fgpyo/util/metric.py

```diff
@@ -245,21 +276,41 @@ def write(cls, path: Path, *values: MetricType) -> None:
         Args:
             path: Path to the output file.
             values: Zero or more metrics.
-
         """
+        header = cls.header()
         with io.to_writer(path) as writer:
             writer.write("\t".join(cls.header()))
             writer.write("\n")
             for value in values:
                 # Important, don't recurse on nested attr classes, instead let implementing class
                 # implement format_value.
-                writer.write("\t".join(cls.format_value(item) for item in value.values()))
+                writer.write(
+                    "\t".join(cls.format_value(item) for item in value.values(_header=header))
+                )
                 writer.write("\n")
 
     @classmethod
+    @final
     def header(cls) -> List[str]:
         """The list of header values for the metric."""
-        return [a.name for a in inspect.get_fields(cls)]  # type: ignore[arg-type]
+        field_types = list(inspect.get_fields(cls))  # type: ignore[arg-type]
+        field_names = {field.name for field in field_types}
+        header = cls._header(field_types=field_types)
+        extra_fields = [h for h in header if h not in field_names]
+        if len(extra_fields) > 0:
+            raise ValueError("header() returned extra fields: " + ", ".join(extra_fields))
+        return header
+
+    @classmethod
+    def _header(cls, field_types: Optional[List[FieldType]] = None) -> List[str]:
```

Comment:
Could I suggest calling this something else - e.g. `_fields_to_write()`?  My reasoning:

1. It's a little confusing to have `header()` and `_header()`
2. More importantly, I _might_ think this would allow me to provide a custom header that doesn't match field names (e.g. `[f.replace("_", " ") for f in in self.header()]`

Reply:
Why is `field_types` optional?  Given this is a function called by the `Metric` class itself, it feels like it should either always be provided, or be removed.  How does the user who overrides this know whether or not they'll get a value?